android/src/BluetoothSensor.java

```diff
@@ -270,48 +281,40 @@ public void onServicesDiscovered(BluetoothGatt gatt,
       return;
     }
 
-    BluetoothGattService service = gatt.getService(BluetoothUuids.HEART_RATE_SERVICE);
-    if (service != null) {
-      BluetoothGattCharacteristic c =
-        service.getCharacteristic(BluetoothUuids.HEART_RATE_MEASUREMENT_CHARACTERISTIC);
-      if (c != null) {
-        setStateSafe(STATE_READY);
-        enableNotification(c);
-      }
-    }
-
-    /* enable notifications for DR Solutions engine sensor box */
-    service = gatt.getService(BluetoothUuids.ENGINE_SENSORS_SERVICE);
-    if (service != null) {
-      BluetoothGattCharacteristic c =
-        service.getCharacteristic(BluetoothUuids.ENGINE_SENSORS_CHARACTERISTIC);
-      if (c != null) {
-        setStateSafe(STATE_READY);
-        enableNotification(c);
-      }
-    }
-
-    /* enable notifications for Flytec Sensbox */
-    service = gatt.getService(BluetoothUuids.FLYTEC_SENSBOX_SERVICE);
-    if (service != null) {
-      BluetoothGattCharacteristic c =
-        service.getCharacteristic(BluetoothUuids.FLYTEC_SENSBOX_NAVIGATION_SENSOR_CHARACTERISTIC);
-      if (c != null) {
-        setStateSafe(STATE_READY);
-        enableNotification(c);
-      }
-
-      c = service.getCharacteristic(BluetoothUuids.FLYTEC_SENSBOX_MOVEMENT_SENSOR_CHARACTERISTIC);
-      if (c != null)
-        enableNotification(c);
-
-      c = service.getCharacteristic(BluetoothUuids.FLYTEC_SENSBOX_SECOND_GPS_CHARACTERISTIC);
-      if (c != null)
-        enableNotification(c);
-
-      c = service.getCharacteristic(BluetoothUuids.FLYTEC_SENSBOX_SYSTEM_CHARACTERISTIC);
-      if (c != null)
-        enableNotification(c);
+    // Check if we know the discovered services & characteristics,
+    // if known, enable notification.
+    // Consecutive calls to getServices() will fail, so we do it just once and
+    // handle the lookups in the loops.
+    List<BluetoothGattService> services = gatt.getServices();
```

Comment:
This will continuously trigger on all service uuids , whether they are supported by xcsoar or not.  Please filter in a list of known UUIDs before continuing.

Reply:
I assumed (backed by observation), that only newly discovered services trigger onServicesDiscovered(). If this is true, the implementation should be okay, right?

Reply:
Added filters for supported services. Now the scanner does only a quick scan for supported services. I am not very happy with the implementation because now the "SERVICE" and "CHARACTERISTIC" from BluetoothUuids.java are mandatory keywords :(
Any suggestions for improvement?

Reply:
I changed BluetoothUuids.java... now I am way more happy with the implementation. The scanner should look only for supported services and after discovering those, the supported characteristics will get notified. Any suggestions?

Reply:
No I think this is good. Ready for merge from my point of view. Unless you have reservations.